metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -1,264 +1,495 @@
 #include "annotated_graph_algorithm.hpp"
 
-#include <unordered_set>
+#include <tsl/hopscotch_set.h>
 
+#include "common/logger.hpp"
+#include "common/seq_tools/reverse_complement.hpp"
 #include "common/vectors/vector_algorithm.hpp"
-#include "common/vector_map.hpp"
-#include "kmer/alphabets.hpp"
-#include "annotation/representation/column_compressed/annotate_column_compressed.hpp"
+#include "common/vectors/bitmap.hpp"
 #include "graph/representation/masked_graph.hpp"
+#include "graph/representation/succinct/dbg_succinct.hpp"
+#include "graph/representation/succinct/boss_construct.hpp"
 
 
 namespace mtg {
 namespace graph {
 
+using mtg::graph::boss::BOSS;
+using mtg::graph::boss::BOSSConstructor;
 using mtg::common::logger;
 
 typedef AnnotatedDBG::node_index node_index;
 typedef AnnotatedDBG::row_index row_index;
 typedef AnnotatedDBG::Annotator::Label Label;
-typedef Alignment<DeBruijnGraph::node_index> DBGAlignment;
 
+typedef std::function<size_t()> LabelCountCallback;
 
-std::unique_ptr<bitmap_vector>
-mask_nodes_by_unitig(const DeBruijnGraph &graph,
-                     const KeepUnitigPath &keep_unitig,
-                     size_t num_threads) {
-    sdsl::bit_vector unitig_mask(graph.max_index() + 1, false);
+constexpr bool MAKE_BOSS = true;
 
-    const bool atomic = num_threads > 1;
 
-    graph.call_unitigs([&](const std::string &unitig, const auto &path) {
-        if (keep_unitig(unitig, path)) {
-            for (DeBruijnGraph::node_index node : path) {
-                set_bit(unitig_mask.data(), node, atomic);
-            }
-        }
-    }, num_threads);
+/**
+ * Return an int_vector<>, bitmap pair, each of length anno_graph.get_graph().get_base_graph().max_index().
+ * For an index i, the int_vector will contain a packed integer representing the
+ * number of labels in labels_in and labels_out which the k-mer of index i is
+ * annotated with. The least significant half of each integer represents the count
+ * from labels_in, while the most significant half represents the count from
+ * labels_out.
+ * The returned bitmap is a binarization of the int_vector
+ */
+std::pair<sdsl::int_vector<>, std::unique_ptr<bitmap>>
+construct_diff_label_count_vector(const AnnotatedDBG &anno_graph,
+                                  const std::vector<Label> &labels_in,
+                                  const std::vector<Label> &labels_out,
+                                  size_t num_threads);
+
+
+template <class GetKeptIntervals>
+void update_masked_graph_by_unitig(MaskedDeBruijnGraph &masked_graph,
+                                   const GetKeptIntervals &get_kept_intervals,
```

Comment:
This is confusing, I had to check out the code and do some search to find the type of `GetKeptIntervals`.
Please add a command explaining this.
Also, maybe pick a better name.
Why is it "kept" and not "masked"?
Why "interval"? Is this an interval of nodes?

Reply:
I added a bunch of comments to describe these.